File: Composite Pattern/Task_Composite.py
import sys
sys.setrecursionlimit(10**9)

class Task:

    def __init__(self,name):
        self.name=name
        self.parent= None

    def get_time_required(self):
        return 0.0


class CompositeTask(Task):
    def __init__(self,name):
        # self.name=name #an alternative for more objective-oriented way of the initialization below
        super().__init__(name)
        self.sub_tasks =[]

    def add_sub_task(self,task):
        self.sub_tasks.append(task)
        task.parent=self

    def get_time_required(self):
        self.time=0.0
        for task in self.sub_tasks:
            self.time = self.time+ task.get_time_required()
        return self.time


class AddDryIngerdientsTask(Task):
    def __init__(self):
        self.name="Add dry ingredients"

    def get_time_required(self):
        return 1.0


class AddLiquidsTask(Task):
    def __init__(self):
        self.name=("Add liquids")

    def get_time_required(self):
        ##demonstrate the usage of non-attribute variable, "time"
        time=0.5
        return time


class MakeBatterTask(CompositeTask):
    def __init__(self):
        self.name=("Make batter")
        self.sub_tasks =[]
        self.add_sub_task(AddDryIngerdientsTask())
        self.add_sub_task(AddLiquidsTask())


class FillPanTask(Task):
    def __init__(self):
        self.name = "Fill the pan!"

    def get_time_required(self):
        return 1.0


class BakeAndFrost(Task):
    def __init__(self):
        self.name= ("Bake the batter and frost it")

    def get_time_required(self):
        return 30.0


class AfterBatterMade(CompositeTask):
    def __init__(self):
        self.name= ("Things after the batter")
        self.sub_tasks = []
        self.add_sub_task(FillPanTask())
        self.add_sub_task(AfterBatterMade())


class MakeCake(CompositeTask):
    def __init__(self):
        self.name= ("Make the whole cake")
        self.sub_tasks = []
        self.add_sub_task(MakeBatterTask())
        self.add_sub_task(AfterBatterMade())

#driver
makeBatter = MakeBatterTask()
print("The time needed to make the baltter is:",makeBatter.get_time_required())

## the following codes cannot be executed due to Python's limitation
makeCake =MakeCake()
print("The time needed to make the whole cake is:",makeCake.get_time_required())

File: Composite Pattern/Task_Composite.rb
class Task
  attr_accessor  :name, :parent

  def initialize(name)
    @name=name
    @parent =nil
  end

  def get_time_required
    0.0
  end
end


class AddDryIngerdientsTask <Task
  def initialize()
    super('Add dry ingredients')
  end

  def get_time_required
    1.0
  end
  end

class AddLiquidsTask <Task
  def initialize()
    super('Add liquids')
  end

  def get_time_required
    0.5
  end
  end


class MixTask <Task
  def initialize()
    super('Mix that batter up!')
  end

  def get_time_required
    3.0
  end
end

class CompositeTask < Task
  def initialize(name)
    super(name)
    @sub_tasks =[]
  end

  def add_sub_task(task)
    # double less than sign "<<" means to append a list in Ruby
    # actually it is also able to be override by new defined semantics to replace "add_sub_task"
    @sub_tasks << task
    task.parent=self
  end

  def remove_sub_task(task)
    @sub_tasks.delete(task)
    task.parent =nil
  end
  def get_time_required
    time =0.0
    @sub_tasks.each { |task|time += task.get_time_required}
    time
  end
end

class MakeBatterTask < CompositeTask
  def initialize
    super("Make batter")
    add_sub_task(AddDryIngerdientsTask.new)
    add_sub_task(AddLiquidsTask.new)
    add_sub_task(MixTask.new)
  end
end

class FillPan <Task
  def initialize()
    super('Fill the pan!')
  end

  def get_time_required
    1.0
  end
end

class BakeAndFrost < Task
  def initialize()
    super('Bake the batter and frost it')
  end

  def get_time_required
    30.0
  end
end


class AfterBatterMade < CompositeTask
  def initialize()
    super("Things after the batter")
    add_sub_task(FillPan.new)
    add_sub_task(BakeAndFrost.new)
  end
end


class MakeCake <CompositeTask
  def initialize
    super ("Make cake")
    add_sub_task(MakeBatterTask.new)
    add_sub_task(AfterBatterMade.new)
  end
end


##driver

composite1 = MakeBatterTask.new
puts("Time required for making the batter is: #{composite1.get_time_required}.")

composite2 = MakeCake.new
puts("Time required for making the whole cake is: #{composite2.get_time_required}.")